core/src/main/java/org/apache/calcite/sql/fun/SqlCaseOperator.java

```diff
@@ -230,14 +232,29 @@ private RelDataType inferTypeFromValidator(
     SqlNodeList thenList = caseCall.getThenOperands();
     ArrayList<SqlNode> nullList = new ArrayList<>();
     List<RelDataType> argTypes = new ArrayList<>();
-    for (SqlNode node : thenList) {
-      argTypes.add(
-          callBinding.getValidator().deriveType(
-              callBinding.getScope(), node));
+
+    final SqlNodeList whenOperands = caseCall.getWhenOperands();
+    final RelDataTypeFactory typeFactory = callBinding.getTypeFactory();
+
+    final int size = thenList.getList().size();
+    for (int i = 0; i < size; i++) {
+      SqlNode node = thenList.get(i);
+      RelDataType type = callBinding.getValidator().deriveType(
+          callBinding.getScope(), node);
+      SqlNode operand = whenOperands.get(i);
+      if (operand.getKind() == SqlKind.IS_NOT_NULL && type.isNullable()) {
+        SqlBasicCall call = (SqlBasicCall) operand;
+        if (call.getOperandList().get(0).equalsDeep(node, Litmus.IGNORE)) {
+          // We're sure that the type is not nullable if the kind is IS NOT NULL.
+          type = typeFactory.createTypeWithNullability(type, false);
+        }
+      }
```

Comment:
The default type created from the type factory is always nullable false, so it make more sense to fix the `IS_NULL` expression.

Reply:
IMO, it won't lose nullable information if there is IS_NULL expression. For instance, 
The type of `CASE WHEN name is null then name else 'NO' end` is:
1) nullable if `name` is nullable
2) not nullable if `name` is not nullable

Am I right?

Reply:
How about if `name` is not nullable and we have a case-branch: `WHEN name is null then name` ? I think this operand should be coerced to nullable(although it is always false).

Reply:
Thanks for your explanation. But in such a case, it seems meaningless to coerce the type to nullable since it never happens. Furthermore, it might also introduce degradation.
So I prefer to not coerce it.

Reply:
I think L266 `typeFactory.leastRestrictive(argTypes)` will make the correct decision to make it nullable.